IDE0057 / "Use range operator": incorrect code change (missing parentheses)

Actual behaviour:
In code like the following:
```
string version = ...;
version = version. Substring(1, (version. IndexOf("-") == -1 ? version. Length : version. IndexOf("-")) - 1)
```
Visual Studio 2019 suggests to &quot;Use range operator&quot;, &quot;IDE0057 Substring can be simplified&quot;, and when invoked, changes the second line to:

```
version = version[1..version. IndexOf("-") == -1 ? version. Length : version. IndexOf("-")];
```
This results in compiler error CS0019: &quot;Operator '==' cannot be applied to operands of type 'Range' and 'int'&quot;, which is correct.

Expected behaviour:
The range operator ``.. `` apparently has higher precedence than the ternary conditional operator ``?:``, so parentheses should have been inserted as appropriate (or rather, in my example, been kept as they were); in this case, the second line might have instead been converted to:
```
version = version[1..( version. IndexOf("-") == -1 ? version. Length : version. IndexOf("-"))];
```

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/633621/ide0057-use-range-operator-incorrect-code-change-m.html
VSTS ticketId: 942430_
_These are the original issue comments:_

Visual Studio Feedback System on 7/5/2019, 00:57 AM (39 min ago): <markdown><p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>
</markdown>
_These are the original issue solutions:_
(no solutions)

--- a/src/Analyzers/CSharp/CodeFixes/UseIndexOrRangeOperator/CSharpUseRangeOperatorCodeFixProvider.cs
+++ b/src/Analyzers/CSharp/CodeFixes/UseIndexOrRangeOperator/CSharpUseRangeOperatorCodeFixProvider.cs
@@ -11,6 +11,7 @@ using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.CodeActions;
 using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Editing;
@@ -177,8 +178,8 @@ namespace Microsoft.CodeAnalysis.CSharp.UseIndexOrRangeOperator
             }
 
             return RangeExpression(
-                startExpr != null && startFromEnd ? IndexExpression(startExpr) : startExpr,
-                endExpr != null && endFromEnd ? IndexExpression(endExpr) : endExpr);
+                startExpr != null && startFromEnd ? IndexExpression(startExpr) : startExpr?.Parenthesize(),
+                endExpr != null && endFromEnd ? IndexExpression(endExpr) : endExpr?.Parenthesize());
         }
 
         private static RangeExpressionSyntax CreateConstantRange(Result result, SyntaxGenerator generator)

--- a/src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryExpressionParenthesesTests.cs
+++ b/src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryExpressionParenthesesTests.cs
@@ -1,4 +1,4 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
+// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
@@ -2488,6 +2488,79 @@ parameters: new TestParameters(options: RemoveAllUnnecessaryParentheses));
         string s2 = f ? """" : default;
     }
 }", offeredWhenRequireForClarityIsEnabled: true);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestRangeWithConstantExpression()
+        {
+            await TestAsync(
+@"class C
+{
+    void M(string s)
+    {
+        _ = s[$$(1)..];
+    }
+}",
+@"class C
+{
+    void M(string s)
+    {
+        _ = s[1..];
+    }
+}", offeredWhenRequireForClarityIsEnabled: true);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestRangeWithMemberAccessExpression()
+        {
+            await TestAsync(
+@"class C
+{
+    void M(string s)
+    {
+        _ = s[$$(s.Length)..];
+    }
+}",
+@"class C
+{
+    void M(string s)
+    {
+        _ = s[s.Length..];
+    }
+}", offeredWhenRequireForClarityIsEnabled: true);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestRangeWithElementAccessExpression()
+        {
+            await TestAsync(
+@"class C
+{
+    void M(string s, int[] indices)
+    {
+        _ = s[$$(indices[0])..];
+    }
+}",
+@"class C
+{
+    void M(string s, int[] indices)
+    {
+        _ = s[indices[0]..];
+    }
+}", offeredWhenRequireForClarityIsEnabled: true);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestRangeWithBinaryExpression()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M(string s)
+    {
+        _ = s[$$(s.Length - 5)..];
+    }
+}", new TestParameters(options: RemoveAllUnnecessaryParentheses));
         }
 
         [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]

--- a/src/Analyzers/CSharp/Tests/UseIndexOrRangeOperator/UseRangeOperatorTests.cs
+++ b/src/Analyzers/CSharp/Tests/UseIndexOrRangeOperator/UseRangeOperatorTests.cs
@@ -781,5 +781,66 @@ class C
                 FixedCode = fixedSource,
             }.RunAsync();
         }
+
+        [WorkItem(36997, "https://github.com/dotnet/roslyn/issues/36997")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseRangeOperator)]
+        public async Task TestExpressionWithAddOperatorArgument()
+        {
+            var source =
+@"
+class C
+{
+    void Goo(string s, int bar)
+    {
+        var v = s.Substring([|bar + 1|]);
+    }
+}";
+            var fixedSource =
+@"
+class C
+{
+    void Goo(string s, int bar)
+    {
+        var v = s[(bar + 1)..];
+    }
+}";
+
+            await new VerifyCS.Test
+            {
+                ReferenceAssemblies = ReferenceAssemblies.NetCore.NetCoreApp31,
+                TestCode = source,
+                FixedCode = fixedSource,
+            }.RunAsync();
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseRangeOperator)]
+        public async Task TestExpressionWithElementAccessShouldNotAddParentheses()
+        {
+            var source =
+@"
+class C
+{
+    void Goo(string s, int[] bar)
+    {
+        _ = s.Substring([|bar[0]|]);
+    }
+}";
+            var fixedSource =
+@"
+class C
+{
+    void Goo(string s, int[] bar)
+    {
+        _ = s[bar[0]..];
+    }
+}";
+
+            await new VerifyCS.Test
+            {
+                ReferenceAssemblies = ReferenceAssemblies.NetCore.NetCoreApp31,
+                TestCode = source,
+                FixedCode = fixedSource,
+            }.RunAsync();
+        }
     }
 }

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ExpressionSyntaxExtensions.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ExpressionSyntaxExtensions.cs
@@ -700,6 +700,12 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
 
                     return OperatorPrecedence.Unary;
 
+                case SyntaxKind.RangeExpression:
+                    // From C# spec, https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/ranges.md#systemrange
+                    // Range: ..
+
+                    return OperatorPrecedence.Range;
+
                 case SyntaxKind.MultiplyExpression:
                 case SyntaxKind.DivideExpression:
                 case SyntaxKind.ModuloExpression:

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/OperatorPrecedence.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/OperatorPrecedence.cs
@@ -23,6 +23,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
         Shift,
         Additive,
         Multiplicative,
+        Range,
         Unary,
         Primary
     }

